src/data/monoid_algebra.lean

```diff
@@ -679,6 +684,42 @@ lemma alg_hom_ext_iff {R : Type u₃} [comm_semiring k] [add_monoid G]
   (∀ x, φ₁ (finsupp.single x 1) = φ₂ (finsupp.single x 1)) ↔ φ₁ = φ₂ :=
 ⟨λ h, alg_hom_ext h, by rintro rfl _; refl⟩
 
+/--
+The `alg_hom` which maps from a grading of an algebra `A` back to that algebra.
+
+The proofs here look an awful lot like `add_monoid_algebra.lift`, but that signature does not allow
+us to distinguish `k` and `A`.
```

Comment:
The general question here is, when A and R are k-algebras, how to describe Hom_{k-Alg}(A[G], R)?
The case A = k is handled by `lift` but in general, when R is noncommutative, the answer isn't so simple. The reason that this `sum_id` map exists is because it's built from the map from G to R = A sending everything to 1, which lands in the center of A.
Noncommutative polynomial evaluation is another special case and I'm not sure we have a totally satisfactory treatment of this special case yet.

Reply:
I started trying to generlalize this and I think ran into the difficulty with commutativity you describe. I think I found that I could replace `b • F a` with `F₁ b * F₂ a` provided I have a hypothesis that `∀ a b, F₁ b * F₂ a = F₂ a * F₁ b. That's at least general enough to recover both `lift` and `sum_id`.

Reply:
Here's what that looks like:
```
variables (k) [comm_semiring k]
variables (G) [monoid G]
variables (A : Type u₃) [semiring A] [algebra k A]
variables (B : Type*) [semiring B] [algebra k B]

def lift_aux (F : G →* B) (H : A →ₐ[k] B) (FH_comm : ∀ b a, H b * F a = F a * H b) : monoid_algebra A G →ₐ[k] B :=
{
  to_fun := λ f, f.sum (λ a b, H b * F a),
  map_one' := by {
    rw [one_def, sum_single_index],
    { rw [F.map_one, H.map_one, one_mul], },
    { rw [F.map_one, H.map_zero, zero_mul], },
  },
  map_mul' := λ f g,
    begin
      rw [mul_def, finsupp.sum_mul, finsupp.sum_sum_index],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      work_on_goal 1 { intros, rw [H.map_add, add_mul], },
      refine finset.sum_congr rfl (λ a ha, _),
      simp only,
      rw [finsupp.mul_sum, finsupp.sum_sum_index],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      work_on_goal 1 { intros, rw [H.map_add, add_mul], },
      refine finset.sum_congr rfl (λ a' ha', _),
      simp only,
      rw [sum_single_index, H.map_mul, F.map_mul],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      calc H (f a) * H (g a') * (F a * F a')
         = H (f a) * (H (g a') * F a) * F a' : by simp only [mul_assoc]
      ...= H (f a) * (F a * H (g a')) * F a' : by rw FH_comm
      ...= H (f a) * F a * (H (g a') * F a') : by simp only [mul_assoc],
    end,
  map_zero' := sum_zero_index,
  map_add' := λ f g,
    begin
      rw [sum_add_index],
      { intros, rw [H.map_zero, zero_mul], },
      { intros, rw [H.map_add, add_mul], },
    end,
  commutes' := λ r,
    begin
      rw [coe_algebra_map, sum_single_index, F.map_one, mul_one, alg_hom.commutes],
      intros, rw [H.map_zero, zero_mul],
    end, }

/-- Any monoid homomorphism `G →* A` can be lifted to an algebra homomorphism
`monoid_algebra k G →ₐ[k] A`. -/
def lift : (G →* A) ≃ (monoid_algebra k G →ₐ[k] A) :=
{ inv_fun := λ f, (f : monoid_algebra k G →* A).comp (of k G),
  to_fun := λ F, lift_aux k G k A F (algebra.of_id k _) (λ a b, by
    simp [algebra.of_id, algebra.algebra_map_eq_smul_one]),
  left_inv := λ f, begin ext x, simp [lift_aux, sum_single_index] end,
  right_inv := λ F,
    begin
      ext f,
      conv_rhs { rw ← f.sum_single },
      simp [algebra.of_id, algebra.algebra_map_eq_smul_one, lift_aux, ← F.map_smul, finsupp.sum, ← F.map_sum],
    end }
```